File: python_lab/problems/01_array_hash_map_basics/001_two_sum.py
"""
Problem 001: 两数之和 (Two Sum)

Difficulty:
Easy

Topic:
Array / Hash Map

Description:
给定一个整数数组 `nums` 和一个整数目标值 `target`，请你在该数组中找出和为目标值 `target` 的那两个整数，并返回它们的数组下标。
你可以假设每种输入只会对应一个答案。但是，数组中同一个元素在答案里不能重复出现。
你可以按任意顺序返回答案。

Input:
- nums: List[int]，整数数组
- target: int，目标和

Output:
- List[int]，包含两个符合条件的下标。

Examples:
Example 1:
Input: nums = [2, 7, 11, 15], target = 9
Output: [0, 1]
Explanation: 因为 nums[0] + nums[1] == 9 ，返回 [0, 1] 。

Example 2:
Input: nums = [3, 2, 4], target = 6
Output: [1, 2]

Constraints:
- 2 <= nums.length <= 10^4
- -10^9 <= nums[i] <= 10^9
- -10^9 <= target <= 10^9

Task:
实现 solution 函数。
"""

from typing import List


def solution(nums: List[int], target: int) -> List[int]:
    # 使用字典记录遍历过的数字及其对应的下标 {数字: 下标}
    seen_map = {}
    
    # 遍历数组，enumerate 会同时返回索引 i 和 值 num
    for i, num in enumerate(nums):
        complement = target - num  # 计算需要的另一半
        
        # 检查另一半是否已经在之前遍历的过程中记录在字典中了
        if complement in seen_map:
            # 如果存在，说明找到了配对！
            # seen_map[complement] 是之前那个数的下标，i 是当前数字的下标
            return [seen_map[complement], i]
        
        # 如果没找到配对，就把当前数字 num 及其下标 i 存入字典中
        seen_map[num] = i
        
    return []


def run_tests():
    assert sorted(solution([2, 7, 11, 15], 9)) == [0, 1], "Test case 1 failed"
    assert sorted(solution([3, 2, 4], 6)) == [1, 2], "Test case 2 failed"
    assert sorted(solution([3, 3], 6)) == [0, 1], "Test case 3 failed"
    print("All test cases passed!")


if __name__ == "__main__":
    run_tests()

File: cpp_lab/problems/01_array_hash_map/001_two_sum.cpp
/*
Problem 001: 两数之和 (Two Sum)

Difficulty:
Easy

Topic:
Array / Hash Map / std::vector / std::unordered_map

Description:
给定一个整数数组 nums 和一个整数 target，请找出和为 target 的两个元素，
并返回它们的下标。

每组输入恰好存在一个答案，同一个元素不能使用两次。
返回的两个下标顺序不限。

Input:
nums: 整数数组
target: 目标和

Output:
包含两个下标的 std::vector<int>

Examples:
Example 1:
Input:
nums = {2, 7, 11, 15}, target = 9

Output:
{0, 1}

Example 2:
Input:
nums = {3, 2, 4}, target = 6

Output:
{1, 2}

Constraints:
- 2 <= nums.size() <= 10^4
- -10^9 <= nums[i] <= 10^9
- -10^9 <= target <= 10^9
- 每组输入恰好存在一个答案

Task:
实现 two_sum 函数。

要求:
1. 目标时间复杂度为 O(n)。
2. 使用 std::unordered_map 记录已经遍历过的数字及其下标。
3. 不允许使用双层循环暴力枚举。
4. 不要修改 nums。
5. 函数参数应避免不必要地复制整个数组。

思考:
- nums 为什么适合通过 const 引用传入？
- unordered_map 应该保存“数字到下标”还是“下标到数字”？
- 为什么应该先查找 complement，再保存当前数字？
- nums 中有重复数字时，怎样避免使用同一个元素两次？
*/

#include <cassert>
#include <cstddef>
#include <iostream>
#include <unordered_map>
#include <vector>


std::vector<int> two_sum(const std::vector<int>& nums, int target) {
    // TODO: 使用一次遍历和 std::unordered_map 完成实现
    return {};
}


void assert_valid_answer(
    const std::vector<int>& nums,
    int target,
    const std::vector<int>& answer
) {
    assert(answer.size() == 2);

    const int first = answer[0];
    const int second = answer[1];

    assert(first >= 0);
    assert(second >= 0);
    assert(static_cast<std::size_t>(first) < nums.size());
    assert(static_cast<std::size_t>(second) < nums.size());
    assert(first != second);
    assert(nums[static_cast<std::size_t>(first)]
        + nums[static_cast<std::size_t>(second)] == target);
}


void run_tests() {
    assert_valid_answer({2, 7, 11, 15}, 9, two_sum({2, 7, 11, 15}, 9));
    assert_valid_answer({3, 2, 4}, 6, two_sum({3, 2, 4}, 6));
    assert_valid_answer({3, 3}, 6, two_sum({3, 3}, 6));
    assert_valid_answer({-3, 4, 3, 90}, 0, two_sum({-3, 4, 3, 90}, 0));
    assert_valid_answer({0, 4, 3, 0}, 0, two_sum({0, 4, 3, 0}, 0));

    std::cout << "All tests passed!\n";
}


int main() {
    run_tests();
    return 0;
}
